Create `#[Serialize]` Attribute to serialize Controller Result

Koc opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 8.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | TBD

`#[Serialize(code: 200, headers: [], context: [])]` Allows automatically serialize Controller Result to format based on Request format.

## Usage example :hammer: 
```php
final readonly class Product
{
    public function __construct(public int $id, public string $name)
    {
    }
}

final readonly class GetProductController
{
    #[Serialize]
    public function __invoke(): Product
    {
        return new Product(1, 'Asus UX550');
    }
}

final readonly class ProductCreated
{
    public function __construct(public int $id)
    {
    }
}

final readonly class CreateProductController
{
    #[Serialize(201)]
    public function __invoke(): ProductCreated
    {
        return new ProductCreated(1);
    }
}

# same controller without annotation: repeatable annoying boilerplate which has no sense to write
final readonly class CreateProductControllerLegacy
{
    public function __construct(private SerializerInterface $serializer)
    {
    }

    #[OA\Response(response: 201, description: 'Successful response', content: new Model(type: ProductCreated::class))]
    public function __invoke(): JsonResponse
    {
        $data = new ProductCreated(1);
        $serialized = $this->serializer->serialize($data, 'json');

        return JsonResponse::fromJsonString($serialized, JsonResponse::HTTP_CREATED);
    }
}
```

## Todo
- [x] add functional tests once #49138 will be merged

Koc committed:
Create `#[Serialize]` Attribute to serialize Controller Result

Koc commented on src/Symfony/Component/HttpKernel/CHANGELOG.md:
```diff
@@ -10,6 +10,7 @@ CHANGELOG
  * Use an instance of `Psr\Clock\ClockInterface` to generate the current date time in `DateTimeValueResolver`
  * Add `#[WithLogLevel]` for defining log levels for exceptions
  * Add `skip_response_headers` to the `HttpCache` options
+ * Add `#[SerializedResponse]` to serialize Controller Result
```
`#[SerializedResponse]` or `#[SerializeResponse]`?

ro0NL replied:
`#[SerializeControllerResult]`?

maybe just `#[Serialize]` even :shrug:

ro0NL commented on src/Symfony/Component/HttpKernel/Attribute/SerializedResponse.php:
```diff
@@ -0,0 +1,23 @@
+#[\Attribute(\Attribute::TARGET_METHOD)]
+class SerializedResponse
+{
+    public function __construct(public readonly int $code = 200, public readonly array $context = [])
```
+ headers?

ro0NL reviewed (commented):
-

ro0NL commented on src/Symfony/Component/HttpKernel/Attribute/Serialize.php:
```diff
@@ -0,0 +1,26 @@
+ * Controller tag to serialize response.
+ */
+#[\Attribute(\Attribute::TARGET_METHOD)]
+class Serialize
```
`AsSerializedResponse` would also work :thinking:

94noni replied:
indeed, others attribute are prefixed with `As` when they describe more the config of the class (AsCommand or AsMessageHandler)
here it is like a mapping/transformation, perhaps like your other PR https://github.com/symfony/symfony/pull/49138 `MapResponse`

94noni commented on src/Symfony/Component/HttpKernel/EventListener/SerializeControllerResultListener.php:
```diff
@@ -0,0 +1,60 @@
+
+final class SerializeControllerResultListener implements EventSubscriberInterface
+{
+    private const DEFAULT_FORMAT = 'json';
```
what is the purpose of such private const?

94noni commented on src/Symfony/Component/HttpKernel/EventListener/SerializeControllerResultListener.php:
```diff
@@ -0,0 +1,60 @@
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            KernelEvents::VIEW => 'onView',
```
how this deals with others listener/subscriber on same event?

ro0NL replied:
you'd overwrite on high prio i suppose, not sure if this listener needs to be low prio

stof commented on src/Symfony/Component/HttpKernel/Attribute/Serialize.php:
```diff
@@ -0,0 +1,26 @@
+ * Controller tag to serialize response.
+ */
+#[\Attribute(\Attribute::TARGET_METHOD)]
+class Serialize
```
this attribute must be final, as the logic in the listener does not account for child classes

Koc replied:
done

stof commented on src/Symfony/Component/HttpKernel/Attribute/Serialize.php:
```diff
@@ -0,0 +1,26 @@
+{
+    public function __construct(
+        public readonly int $code = 200,
+        public readonly array $headers = [],
```
you should add phpdoc describing the structure of those arrays (`$headers` is probably `array<string, string>` for instance)

Koc replied:
ok, but why Psalm not violates about this?

Koc replied:
done

stof replied:
Probably because it is not configure to require providing this info.

stof commented on src/Symfony/Component/HttpKernel/EventListener/SerializeControllerResultListener.php:
```diff
@@ -0,0 +1,60 @@
+
+        $request = $event->getRequest();
+        $controllerResult = $event->getControllerResult();
+        $format = $request->getPreferredFormat(self::DEFAULT_FORMAT);
```
This listener should rather use `$request->getRequestFormat()`, getting the format configured for the response to that request, instead of enforcing usage of negotiation. This would make it compatible with all ways to determine the response format.

Koc replied:
done

yceruto commented:
This test is failing https://github.com/symfony/symfony/actions/runs/4272061115/jobs/7436960556#step:7:8239

You probably have to add `symfony/serializer` as a dev dependency of the http-kernel component.

Koc commented:
yes, thanx, it was added in #49138. I've not added it here to prevent merge conflicts. Also I want extend functional tests added in mentioned PR with new Attribute

ro0NL reviewed (commented):
-

nicolas-grekas commented:
For what it's worth, I'm personally not convinced we should merge this. The reason is that the resulting behavior is very limited and resorting to explicit code comes very quickly - as soon as one needs more precise control of the resulting response for example. In the end, this saves two/three lines that are legit to have in a controller/similar to me.

Koc commented:
Exactly, this can save few lines in the Controller and reduce boilerplate. Of course it can't cover all needs for an every project, but same applicable for other stuff provided by Framework. 80/20 rule in action. We have few microservices that uses only `#[Serialize]` Attribute without any kind of additional control for the Response.

Koc commented:
Rebased, renamed `context` to `serializationContext` for consistency for `Map*` attributes. Test failures looks unrelated.

@nicolas-grekas please review :pray:

Koc commented:
I've rebased PR one more time + fixed tests + actualized changelog.

And as I've already mentioned - this feature needed not only for reducing boilerplate of serialization but also for simplification of the OpenApi spec generation via NelmioApiDocBundle. They already [successfully adopted](https://symfony.com/bundles/NelmioApiDocBundle/current/symfony_attributes.html) `#[MapRequestPayload/MapQueryString]`. So, introduction of this attribute can be a good step forward in same direction.

DavidGarciaCat commented on src/Symfony/Component/HttpKernel/Attribute/Serialize.php:
```diff
@@ -0,0 +1,34 @@
+ * @author Konstantin Myakshin <molodchick@gmail.com>
+ */
+#[\Attribute(\Attribute::TARGET_METHOD)]
+final class Serialize
```
Shouldn't this `#[Serialize]` attribute allow to pass the groups to be used during the serialization process?

Also, the `SerializerInterface` and its `serialize` method allows us to pass the preferred format (JSON, XML, ...) so making it an argument (even with a predefined value) makes sense to me 🤔 

==== EDIT ====

By the way: with "passing groups" I meant that they seem to be a relevant setting so maybe it's a good idea to add them with a shortcut (instead of passing them via context), but that's just a secondary matter - the primary concern here is the missing format:

```php
$data = $this->serializer->serialize(
            data: $output,
            format: 'json',
            context: ['groups' => ['Group1', 'Group2']]
        );
```

so it could be like this:

```php
#[Serialize(code: 201, format: 'json', headers: [], serializationContext: ['groups' => ['Group1', 'Group2']])]
```

or even

```php
#[Serialize(code: 201, format: 'json', headers: [], groups: ['Group1', 'Group2'])]
```

rrajkomar commented on src/Symfony/Component/HttpKernel/Attribute/Serialize.php:
```diff
@@ -0,0 +1,34 @@
+ * @author Konstantin Myakshin <molodchick@gmail.com>
+ */
+#[\Attribute(\Attribute::TARGET_METHOD)]
+final class Serialize
```
Hi,
I think this is a great idea and would be very useful for anyone developing webservice so thank you.
However wouldn't it be more logical for consistency to name the attribute in the same manner as the others ?
Something like `MapResponsePayload` seems more intuitive IMHO.

Koc replied:
this attributes can be renamed soon, see #58709

Koc commented:
@nicolas-grekas I've updated current PR to take into account your recent changes from #63032.

In general most of the pipelines pass (there is a failed unrelated test in Console suite). But [few](https://github.com/symfony/symfony/actions/runs/21731286342/job/62686243373?pr=49518) of them are failed with an error:

```
1) Symfony\Component\HttpKernel\Tests\EventListener\SerializeControllerResultListenerTest::testSerializeAttribute
TypeError: Symfony\Component\HttpKernel\EventListener\SerializeControllerResultAttributeListener::onView(): 
Argument #1 ($event) must be of type Symfony\Component\HttpKernel\Event\ControllerAttributeEvent, 
Symfony\Component\HttpKernel\Event\ViewEvent given, called in /home/runner/work/symfony/symfony/src/Symfony/Component/HttpKernel/Tests/EventListener/SerializeControllerResultListenerTest.php on line 54

/home/runner/work/symfony/symfony/src/Symfony/Component/HttpKernel/EventListener/SerializeControllerResultAttributeListener.php:31
/home/runner/work/symfony/symfony/src/Symfony/Component/HttpKernel/Tests/EventListener/SerializeControllerResultListenerTest.php:54
/home/runner/work/symfony/symfony/.phpunit/phpunit-12-0/phpunit:104
/home/runner/work/symfony/symfony/vendor/symfony/phpunit-bridge/bin/simple-phpunit.php:458
/home/runner/work/symfony/symfony/vendor/symfony/phpunit-bridge/bin/simple-phpunit:13
```

Should I apply same stuff line in other attribute listeners `if (!class_exists(ControllerAttributesListener::class, false)) {`? I thought that if my listeners was introduced in the version 8.1 - no need to add such workarounds as `ControllerAttributeEvent` will be always available :thinking: .

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/composer.json:
```diff
@@ -28,7 +28,7 @@
         "symfony/finder": "^7.4|^8.0",
         "symfony/http-foundation": "^7.4|^8.0",
-        "symfony/http-kernel": "^7.4|^8.0",
+        "symfony/http-kernel": "^8.1",
```
instead of bumping, let's remove the service in FrameworkExtension

Koc replied:
done

nicolas-grekas commented on src/Symfony/Component/HttpKernel/Attribute/Serialize.php:
```diff
@@ -0,0 +1,33 @@
+    public function __construct(
+        public readonly int $code = 200,
+        public readonly array $headers = [],
+        public readonly array $serializationContext = [],
```
maybe simplify?
```suggestion
        public readonly array $context = [],
```

Koc replied:
done

nicolas-grekas commented on src/Symfony/Component/HttpKernel/CHANGELOG.md:
```diff
@@ -19,6 +19,7 @@ CHANGELOG
  * Deprecate the `Symfony\Component\HttpKernel\DependencyInjection\Extension` class, use the parent `Symfony\Component\DependencyInjection\Extension\Extension` class instead
  * Allow using Expression or \Closure for `validationGroups` in `#[MapRequestPayload]` and `#[MapQueryString]`
  * Deprecate passing a `ControllerArgumentsEvent` to the `ViewEvent` constructor; pass a `ControllerArgumentsMetadata` instead
+ * Add `#[Serialize]` to serialize Controller Result
```
```suggestion
 * Add `#[Serialize]` to serialize values returned by controllers
```

Koc replied:
done

nicolas-grekas commented on src/Symfony/Component/HttpKernel/EventListener/SerializeControllerResultAttributeListener.php:
```diff
@@ -0,0 +1,64 @@
+            return;
+        }
+
+        /** @var Serialize $attribute */
```
instead: `@param ControllerAttributeEvent<Serialize> $event` on the method

Koc replied:
done

nicolas-grekas commented on src/Symfony/Component/HttpKernel/EventListener/SerializeControllerResultAttributeListener.php:
```diff
@@ -0,0 +1,64 @@
+        $attribute = $event->attribute;
+
+        if (!$this->serializer) {
+            throw new \LogicException(\sprintf('The "symfony/serializer" component is required to use the "%s" attribute. Try running "composer require symfony/serializer".', Serialize::class));
```
```suggestion
            throw new \LogicException(\sprintf('The "symfony/serializer" component is required to use the "#[%s]" attribute. Try running "composer require symfony/serializer".', Serialize::class));
```

Koc replied:
done

nicolas-grekas commented on src/Symfony/Component/HttpKernel/EventListener/SerializeControllerResultAttributeListener.php:
```diff
@@ -0,0 +1,64 @@
+        $controllerResult = $kernelEvent->getControllerResult();
+        $format = $request->getRequestFormat('json');
+
+        $data = $this->serializer->serialize($controllerResult, $format, $attribute->serializationContext);
```
what's going to be the behavior in case of unsupported format? let's figure out the proper HTTP response code?

Koc replied:
done, please check

nicolas-grekas commented on src/Symfony/Component/HttpKernel/EventListener/SerializeControllerResultAttributeListener.php:
```diff
@@ -0,0 +1,64 @@
+
+        $data = $this->serializer->serialize($controllerResult, $format, $attribute->serializationContext);
+
+        $headers = $attribute->headers + ['Content-Type' => $request->getMimeType($format)];
```
let's not call getMimeType is the header is already set in the attribute? also, the current logic is case-sensitive, this can give weird results

Koc replied:
done, please check

nicolas-grekas reviewed (commented):
Nice. Some nitpicking and should be ready on my side.

Koc commented:
Test failures in console component looks unrelated :face_with_diagonal_mouth:

nicolas-grekas commented:
Thank you @Koc.
